### PR TITLE
Remove Held Write Acks

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR removes held write acks, while mostly maintaining the existing user-facing properties. Instead of keeping mutations in the mutation queue, they are directly applied to the RemoteDocumentCache. As part of this, the cache entry now tracks the "commit version" and we can thus tell whether a document is dirty.

If a document is mutated after query initialization time (the first SnapshotVersion received from the backend), we raise  `hasPendingWrites`. For documents that were modified before the query was registered, we raise `isFromCache` (this is meant to mirror the previous behavior, as we cleared write acks for no listen).

There is special logic to that deals with patches to unknown documents. Updates to these documents get filtered in the view until Watch catches up.

This is a large PR that rewrites some of our core view handling, as as such, it is likely beneficial to split this up. I am sending this out in one big chunk for now to get some high level feedback. Once that has been addressed, I plan to split this PR up.

Possible splitting points:
- Extract the schema migration that drops held write acks. This will also include the plumbing of DatabaseId to `createOrUpgradeDb`.
- Changing of `removeMutationBatches` to only accept a single batch and dropping of `getHighestAcknowledgedBatchID`.
- Addition of `SnapshotVersion` to SharedClientState (used to update the Views of secondary tabs)
- The rest of the PR, including the new and updated Spec tests.

There are also some more unit tests that I should add for the View changes.

### CODE REVIEW

File: packages/firestore/src/local/shared_client_state.ts [hunk 1]
```diff
@@ -173,6 +176,7 @@ export interface SharedClientState {
  * encoded as part of the key.
  */
 interface MutationMetadataSchema {
+  snapshotVersionMicros: number;
```
Comment (schmidt-sebastian) [inline packages/firestore/src/local/shared_client_state.ts]: We should store this as seconds/nanos.

File: packages/firestore/src/core/view.ts [hunk 2]
```diff
@@ -82,6 +90,18 @@ export class View {
   private limboDocuments = documentKeySet();
   /** Document Keys that have local changes */
   private mutatedKeys = documentKeySet();
+  /**
+   * The first snapshot version that Watch has sent us during the lifetime of
+   * of this view. Any document edited after will have `hasPendingWrites` set
+   * until Watch catches up.
+   */
+  private firstSnapshotVersion = SnapshotVersion.MIN;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: I removed these snapshot versions:
- Since we store UnknownDocuments in the database, we no longer need to check whether the snapshots "went back in time".
- I added a special `computeInitialChanges()` method that gets rid of the need for `firstSnapshotVersion`. This method differs from computeDocChanges in that ignores `hasCommittedMutations` for the first view snapshot.

File: packages/firestore/src/core/view.ts [hunk 3]
```diff
@@ -82,6 +90,18 @@ export class View {
...
+  /**
+   * The last snapshot version that this view has been synced with. Used to
+   * ignore events if the Write and the Watch stream send us update out of
+   * order.
+   */
+  private lastSnapshotVersion = SnapshotVersion.MIN;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: This comment should have been resolved by the changes in this and the earlier PRs.

File: packages/firestore/src/core/view.ts [hunk 4]
```diff
@@ -105,12 +125,15 @@ export class View {
    * need to go back to the local cache for more results. Does not make any
    * changes to the view.
    * @param docChanges The doc changes to apply to this view.
+   * @param snapshotVersion The snapshot version that triggered the doc changes
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: This argument has been removed.

File: packages/firestore/src/local/local_store.ts [hunk 5]
```diff
@@ -558,27 +473,50 @@ export class LocalStore {
...
           documentBuffer.getEntry(txn, key).next(existingDoc => {
+            let shouldApplyUpdate = false;
+
             // If a document update isn't authoritative, make sure we don't
             // apply an old document version to the remote cache. We make an
             // exception for SnapshotVersion.MIN which can happen for
             // manufactured events (e.g. in the case of a limbo document
             // resolution failing).
+
+            if (authoritativeUpdates.has(doc.key)) {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/local_store.ts]: I wanted to be able to flip `shouldApplyUpdate` back to false, and that's how I ended up changing these statements.

In the end, most of these changes are no longer needed so I reverted almost all of this.

File: packages/firestore/src/local/memory_remote_document_cache.ts [hunk 6]
```diff
@@ -91,7 +91,11 @@ export class MemoryRemoteDocumentCache implements RemoteDocumentCache {
       changedDocs = changedDocs.insert(
         key,
         this.docs.get(key) ||
-          new NoDocument(key, SnapshotVersion.forDeletedDoc())
+          new NoDocument(
+            key,
+            SnapshotVersion.forDeletedDoc(),
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/memory_remote_document_cache.ts]: All documents now only have one version.

File: packages/firestore/src/local/shared_client_state.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/src/local/shared_client_state.ts]: No longer applicable since we don't need to look these version up from LocalStorage.

File: packages/firestore/src/local/simple_db.ts [hunk 7]
```diff
@@ -35,10 +36,12 @@ const LOG_TAG = 'SimpleDb';
 export class SimpleDb {
   /** Opens the specified database, creating or upgrading it if necessary. */
   static openOrCreate(
+    databaseId: DatabaseId,
     name: string,
     version: number,
     runUpgrade: (
       db: IDBDatabase,
+      databaseId: DatabaseId,
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/simple_db.ts]: This was used in the serializer. Addressed in earlier PR.

File: packages/firestore/src/model/document.ts [hunk 8]
```diff
@@ -51,19 +52,38 @@ export class Document {
     return this.data.value();
   }
 
+  hasPendingWrites(baseSnapshotVersion: SnapshotVersion): boolean {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/document.ts]: This is much simpler now: A Document has pending writes when either localMutations or committedMutations are present. A NoDocument has pending writes when committedMutations (note that this is internal only and not exposed to the user).

File: packages/firestore/src/model/document.ts [hunk 9]
```diff
@@ -51,19 +52,38 @@ export class Document {
...
+    if (this.hasLocalMutations) {
+      return true;
+    }
+
+    // If the document was committed after Watch has delivered the base
+    // snapshot, we raise `hasPendingWrites` as long as the local commit version
+    // remains higher than the version sent to us by Watch.
+    if (
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/document.ts]: No longer applicable.

File: packages/firestore/src/model/document.ts [hunk 10]
```diff
@@ -89,21 +109,26 @@ export class Document {
  * denotes time we know it didn't exist at.
  */
 export class NoDocument {
-  constructor(readonly key: DocumentKey, readonly version: SnapshotVersion) {}
+  constructor(
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/document.ts]: This has been resolved in an earlier PR.

File: packages/firestore/test/unit/api/database.test.ts [hunk 11]
```diff
@@ -80,32 +80,39 @@ describe('Query', () => {
 describe('QuerySnapshot', () => {
   it('support equality checking with isEqual()', () => {
     expectEqual(
-      querySnapshot('foo', {}, { a: { a: 1 } }, true, false, false),
-      querySnapshot('foo', {}, { a: { a: 1 } }, true, false, false)
+      querySnapshot('foo', {}, { a: { a: 1 } }, keys('coll/a'), false, false),
```
Comment (wilhuff) [inline packages/firestore/test/unit/api/database.test.ts]: "coll/a" seems to be unrelated to the collection in question.

Also, it's weird that this can have mutated keys but neither from cache nor sync state changed flags set.

File: packages/firestore/src/core/transaction.ts [hunk 12]
```diff
@@ -36,7 +37,7 @@ export class Transaction {
 
   constructor(private datastore: Datastore) {}
 
-  private recordVersion(doc: MaybeDocument): void {
+  private recordVersion(doc: NoDocument | Document): void {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/transaction.ts]: I already changed this when I did the initial Android port. This function takes a MaybeDocument and has an assert inside of it. The same is now true for the Web.

File: packages/firestore/src/core/view.ts [hunk 13]
```diff
@@ -99,6 +99,58 @@ export class View {
...
+   * Unlike `computeDocChanges`, documents with committed mutations don't raise
+   * `hasPendingWrites`. This distinction allows us to only raise
+   * `hasPendingWrite` events for documents that changed during the lifetime of
+   * the View.
+   *
+   * @param docs The docs to apply to this view.
+   * @return A new set of docs, changes, and refill flag.
+   */
+  computeInitialChanges(docs: MaybeDocumentMap): ViewDocumentChanges {
+    // TODO: Call this directly from the constructor and use
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: Removed the TODO for now. The initial setup to get both a ViewSnapshot and a View involves a lot of back-and-forth of the same data, and I was leaning towards reducing this. I'll think some more about this after all these PRs are merged.

File: packages/firestore/src/core/view.ts [hunk 14]
```diff
@@ -153,53 +205,72 @@ export class View {
...
 
-            if (
-              lastDocInLimit &&
-              this.query.docComparator(newDoc, lastDocInLimit) > 0
-            ) {
-              // This doc moved from inside the limit to after the limit.
-              // That means there may be some doc in the local cache that's
-              // actually less than this one.
-              needsRefill = true;
+              if (
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: It's simpler now to do this check only in this branch (since we otherwise also have to include `changeApplied` in the if-check). I can't think of a case where a document would fall outside the limit based on a metadata-only change.

File: packages/firestore/src/core/view.ts [hunk 15]
```diff
@@ -223,6 +294,23 @@ export class View {
...
+    // We suppress the initial change event for documents that were
+    // modified as part of a write acknowledgment (e.g. when the value of
+    // a server transform is applied) as Watch will send us the same
+    // document again.
+    // By suppressing the event, we only raise two user visible events
+    // (one with `hasPendingWrites` and the final state of the document)
+    // instead of three (one with `hasPendingWrites`, the modified
+    // document with `hasPendingWrites` and the final state of the
+    // document).
+    return (
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/core/view.ts]: I used to have a version check in here when I still dealt with versions in the view. Now that I have the UnknownDocument, I no longer need to look at the document version, but only on the committed flag. If I know that a document had mutations committed, then I want to wait for the acknowledgement from Watch. 

BTW, this essentially ignores all documents updated in a WriteAck and we could look at handling this outside the View.

I renamed this method to `shouldWaitForSyncedDocument`. I debated naming 'synced' 'remote' but that term is a little bit overloaded already in our SDK.

File: packages/firestore/src/local/local_serializer.ts [hunk 16]
```diff
@@ -56,7 +56,9 @@ export class LocalSerializer {
     } else if (remoteDoc.noDocument) {
       const key = DocumentKey.fromSegments(remoteDoc.noDocument.path);
       const version = this.fromDbTimestamp(remoteDoc.noDocument.readTime);
-      return new NoDocument(key, version);
+      return new NoDocument(key, version, {
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/local/local_serializer.ts]: DbRemoteDocument is the persisted version of the document. I felt that it was saner to directly encode the boolean flag there, especially since the schema is already in JSON and has the name encoded.

DbNoDocument looks this way to match DbDocument.

File: packages/firestore/test/unit/api/database.test.ts (hunk 11 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/unit/api/database.test.ts]: I changed these tests to use `foo`.

I also went through and removed the mutations from most of these comparisons. Only the one case that directly compared the mutated keys still has mutated keys with both flags false (to make it obvious what is tested here). 
BTW, I think this case could happen, albeit it would be rare (the backend has sent us all committed documents, but we are still not CURRENT since there are more edits from other clients).

I also fixed the ViewSnapshot comparison to take `mutatedKeys` into account.

File: packages/firestore/test/util/api_helpers.ts [hunk 17]
```diff
@@ -87,7 +95,7 @@ export function query(path: string): Query {
  *     document, with the key being the document id, and the value being the document contents.
  * @param docsToAdd Specifies data to be added into the query snapshot as of now. Each entry maps
  *     to a document, with the key being the document id, and the value being the document contents.
- * @param hasPendingWrites Whether the query snapshot has pending writes.
+ * @param mutatedKes The list of document with pending writes.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/util/api_helpers.ts]: Fixed.
